File: Solutions/Main/NhsCui.Toolkit.Web.ClientTest/DateTest.js
//-----------------------------------------------------------------------
// <copyright file="DateTest.js" company="Microsoft Corporation and Crown copyright 2007 - 2010.">
// Copyright (c) Microsoft Corporation and Crown copyright 2007 - 2010.
// All rights reserved.
//
// CERTAIN PARTS OF THIS WORK CONTAIN SOFTWARE CODE THAT IS LICENSED
// FOR USE UNDER THE MICROSOFT PUBLIC LICENSE. DISTRIBUTION, IN SOURCE CODE
// OR OBJECT CODE FORM, OF THOSE PARTS MUST COMPLY WITH THE TERMS OF THE
// PUBLIC LICENSE. SEE http://www.microsoft.com/opensource/licenses.mspx
// FOR DETAILS.
// IF YOU BRING A PATENT CLAIM AGAINST ANY CONTRIBUTOR OVER PATENTS THAT
// YOU CLAIM ARE INFRINGED BY THE PUBLIC LICENSE SOFTWARE, YOUR PATENT
// LICENSE FROM SUCH CONTRIBUTOR TO THE SOFTWARE ENDS AUTOMATICALLY.
//
// THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY 
// KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
// PARTICULAR PURPOSE.
// </copyright>
// <date>03-Jan-2007</date>
// <summary>Client-side javascript for NhsDate tests</summary>
//-----------------------------------------------------------------------

Type.registerNamespace("NhsCui.Toolkit.Web.Tests");

/// <summary>
/// Unit tests to test the Time class in the NhsCui.Toolkit.DateAndTime namespace
/// </summary>
var NhsDateTest = NhsCui.Toolkit.Web.Tests.NhsDateTest = function() {
}

NhsCui.Toolkit.Web.Tests.NhsDateTest.prototype = {

    /// <summary>
    /// Test the ArgumentNullException for instruction parameter
    /// </summary>
//    [ExpectedException(typeof(System.ArgumentNullException), "Should throw Null argument exception from Add method")]
    NullInstructionAddTest : function()
    {
        var nowAsItWasAtTheStartofTheTest = new Date();
        var date = new NhsDate(nowAsItWasAtTheStartofTheTest);
        NhsDate.add(date, null);            
    },

    /// <summary>
    /// Test the ArgumentNullException for sourcedate parameter
    /// </summary>
//    [ExpectedException(typeof(System.ArgumentNullException), "Should throw Null argument exception from Add method")]
    NullSourceDateAddTest : function()
    {
        NhsDate.add(null, String.format("+1{0}", NhsDateResources.DaysUnit));
    },
    
    /// <summary>
    /// Test the ArgumentOutOfRangeException for instruction parameter
    /// </summary>
//    [ExpectedException(typeof(System.ArgumentOutOfRangeException), "Should throw ArgumentOutOfRangeException from Add method")]
    InvalidInstructionAddTest : function()
    {
        var date = new Date();
        var sourcedate = new NhsDate(date);
        NhsDate.add(sourcedate, String.format("+1{0}", "X"));
    },

    /// <summary>
    /// Test the ArgumentOutOfRangeException for sourcedate parameter being a year
    /// </summary>
//    [ExpectedException(typeof(System.ArgumentOutOfRangeException), "Should throw ArgumentOutOfRangeException from Add method")]
    SourceDateYearAddTest : function()
    {
        var year = 2007;
        var date = new NhsDate(year);
        NhsDate.add(date, String.format("+1{0}", NhsDateResources.DaysUnit));
    },

    /// <summary>
    /// Test the ability to do simple arithemtic updates to date by asking Date to add 1 day
    /// </summary>
    DataAddDay : function()
    {
        var nowAsItWasAtTheStartofTheTest = new Date(); 

        var date = new NhsDate(nowAsItWasAtTheStartofTheTest);

        date.add(String.format("+1{0}", NhsDateResources.DaysUnit));

        nowAsItWasAtTheStartofTheTest.setDate(nowAsItWasAtTheStartofTheTest.getDate() + 1)
        
        Assert.AreEqual(date.get_dateValue().valueOf(), nowAsItWasAtTheStartofTheTest.valueOf(), "Dates should match");
    },

    /// <summary>
    /// Test the ability to do complex arithemtic updates to date by asking Date to add 1 month and 1 day
    /// </summary>
    DataAddMonthAndDay : function()
    {
        var dateToAddTo = new Date(2007,0,29); 

        var date = new NhsDate(dateToAddTo);

        date.add(String.format("+1{0}+1{1}", NhsDateResources.MonthsUnit, NhsDateResources.DaysUnit));
        
        Assert.AreEqual(new Date(2007,2,1).valueOf(), date.get_dateValue().valueOf(), "Dates should match");
    },

    /// <summary>
    /// Test the ability to do simple arithemtic updates to date by asking Date to Add 7 years
    /// </summary>
    DataAdd7Years : function()
    {
        var dateToAddTo = new Date(2001,1,2); 

        var date = new NhsDate(dateToAddTo);

        date.add(String.format("+7{0}", NhsDateResources.YearsUnit));

        //Sys.Debug.trace("DataAdd7Years: ACTUAL " + date.get_dateValue().valueOf())
                
        Assert.AreEqual(new Date(2008,1,2).valueOf(), date.get_dateValue().valueOf(), "Dates should match");
    },

    /// <summary>
    /// Test the ability to do simple arithemtic updates to date by asking Date to Subtract 7 years
    /// </summary>
    DataSubtract7Years : function()
    {
        var dateToSubtractFrom = new Date(2009,3,4); 

        var date = new NhsDate(dateToSubtractFrom);

        date.add(String.format("-7{0}", NhsDateResources.YearsUnit));
        
        //Sys.Debug.trace("DataSubtract7Years: ACTUAL " + date.get_dateValue().valueOf())
        
        Assert.AreEqual(new Date(2002,3,4).valueOf(), date.get_dateValue().valueOf(), "Dates should match");
    },

    /// <summary>
    /// Test the ability to do complex arithemtic updates to date by asking Date to Subtract 2 years AND add 5 days
    /// </summary>
    DataSubtract2YearsAdd5Days : function()
    {
        var dateToAddTo = new Date(2004, 1, 28); 

        var date = new NhsDate(dateToAddTo);

        date.add(String.format("-2{0}+5{1}", NhsDateResources.YearsUnit, NhsDateResources.DaysUnit));
        
        Sys.Debug.trace("DataSubtract2YearsAdd5Days: ACTUAL " + date.get_dateValue().valueOf())
        
        Assert.AreEqual(new Date(2002, 2, 5).valueOf(), date.get_dateValue().valueOf(), "Dates should match");
    },

    /// <summary>
    /// Test ToString on DateType Exact
    /// </summary>
    ToStringExact : function()
    {
        var date = new NhsDate(new Date(1974, 2, 26));

        Assert.AreEqual("26-Mar-1974", date.toString(), "Exact date should match");

        Assert.AreEqual("26-Mar-1974", date.toString(false), "Check that default for 'include day of week' flag is correct");

        // try with includeDayOfWeek flag
        Assert.AreEqual("Tue 26-Mar-1974", date.toString(true), "Exact date with DayOfWeek should match");
    },

    /// <summary>
    /// Test ToString on DateType Exact
    /// </summary>
    ToStringExactGetRelativeText : function()
    {
        var date = new NhsDate(new Date());

        var gc = new GlobalizationService();

        // Today works
        Assert.AreEqual(NhsDateResources.Today, date.toString(false, false, true), "RelativeText of Today is not working");

        date.get_dateValue().setDate(date.get_dateValue().getDate()+1);

        Assert.AreEqual(NhsDateResources.Tomorrow, date.toString(false, false, true), "RelativeText of Tomorrow is not working");

        date.get_dateValue().setDate(date.get_dateValue().getDate()-2);

        Assert.AreEqual(NhsDateResources.Yesterday, date.toString(false, false, true), "RelativeText of Yesterday is not working");
        
        // Check the "Who Wins condition when "Show Day of Week" and "Show Relative Text" are both true
        Assert.AreEqual(NhsDateResources.Yesterday, date.toString(true, false, true), "When includeDayOfWeek, showRelativeText are both true, showRelativeText should win. It is not");

        var testDate = new Date(1974, 3, 26);

        date.set_dateValue(testDate);

        Assert.AreEqual(testDate.format(gc.shortDatePattern), date.toString(), "Check that default for 'include day of week' flag is correct");

        // try with includeDayOfWeek flag

        Assert.AreEqual(testDate.format(gc.shortDatePatternWithDayOfWeek), date.toString(true), "Check that default for 'include day of week' flag is correct");
    },

    /// <summary>
    /// Test ToString on DateType Year
    /// </summary>
    ToStringYear : function()
    {
        var baseDateTime = new Date();

        var date = new NhsDate(baseDateTime);

        date.set_dateType(DateType.Year);

        Assert.AreEqual("0000", date.toString(), "Check the default of Year when an explicit date is passed to ctor");

        date = new NhsDate(baseDateTime.getYear());

        date.set_dateType(DateType.Year);

        Assert.AreEqual(baseDateTime.getYear().toString(), date.toString());
    },

    /// <summary>
    /// Test ToString on DateType YearMonth
    /// </summary>
    ToStringYearMonth : function()
    {
        var baseDateTime = new Date();

        var date = new NhsDate(baseDateTime);

        date.set_dateType(DateType.YearMonth);

        Assert.AreEqual(String.format("{0}-{1}", Sys.CultureInfo.CurrentCulture.dateTimeFormat.MonthNames[1 - 1], "0000"), date.toString(), "Check the default of Month and Year when an explicit date is passed to ctor");
        
        // Now set the date to a year and a month which is what DateType = YearMonth prefers
        date = new NhsDate(1974, 3);

        Assert.AreEqual(
            String.format("{0}-{1}", Sys.CultureInfo.CurrentCulture.dateTimeFormat.MonthNames[3 - 1], "1974"), date.toString());
    },

    /// <summary>
    /// Test ToString on DateType Null
    /// </summary>
    ToStringNull : function()
    {
        var date = new NhsDate();
        date.set_dateType(DateType.Null);
        Assert.IsTrue(date.toString().length === 0, "Expecting ToString() on null date to return emptystring");
    },

    /// <summary>
    /// Test Parse for a Month and Year
    /// </summary>
    ParseMonthAndYear : function()
    {
        var baseDateTime = new Date();

        var date = new NhsDate(baseDateTime);

        date.set_dateType(DateType.YearMonth);
        date.set_year(1974);
        date.set_month(3);

        var parsedDate;

        Assert.IsTrue((parsedDate=NhsDate.tryParse(date.toString()))!=null, String.format("TryParse failed with date, {0}", date.toString()));

        Assert.AreEqual(date.get_month(), parsedDate.get_month(), "Parse did not error but Month does not match");
        
        Assert.AreEqual(date.get_dateType(), parsedDate.get_dateType(), "Parse did not error but DateType does not match");
        Assert.AreEqual(date.get_year(), parsedDate.get_year(), "Parse did not error but Year does not match");
    },

    /// <summary>
    /// Test Parse for a Null
    /// </summary>
    ParseNullIndex : function()
    {
        var date = new NhsDate();

        date.set_dateType(DateType.NullIndex);
        date.set_nullIndex(4);
        
        // Muddy the water by assigning Month and year
        date.set_year(2007);
        date.set_month(3);

        var parsedDate;

        Assert.IsTrue((parsedDate=NhsDate.tryParse(date.toString()))!=null, String.format("TryParse failed with date, {0}", date.toString()));

        Assert.AreEqual(date.get_dateType(), parsedDate.get_dateType(), "Parse did not error but DateType does not match");
        Assert.AreEqual(date.get_nullIndex(), parsedDate.get_nullIndex(), "Parse did not error but NullIndex does not match");
    },

    /// <summary>
    /// Test Parse for a free text
    /// </summary>
    ParseDateText : function()
    {
        var date = new NhsDate("26-Mar-1974");

        var parsedDate;

        Assert.IsTrue((parsedDate=NhsDate.tryParse(date.toString()))!=null, String.format("TryParse failed with date, {0}", date.toString()));

        Assert.AreEqual(date.get_dateType(), parsedDate.get_dateType(), "Parse did not error but DateType does not match");
        Assert.AreEqual(date.get_dateValue().toString(), parsedDate.get_dateValue().toString(), "Parse did not error but DateValue does not match");
    },
    
    /// <summary>
    /// Test Parse null
    ///</summary>
    ParseNull : function()
    {
        var date = NhsDate.parse(null);
        Assert.AreEqual(date.get_dateType(), DateType.Null, "Parse did not error but DateType does not match");
    },

    /// <summary>
    /// Test Parse empty string
    ///</summary>
    ParseEmptyString : function()
    {
        var date = NhsDate.parse("");
        Assert.AreEqual(date.get_dateType(), DateType.Null, "Parse did not error but DateType does not match");
    },

    /// <summary>
    /// Test IsNull Property
    ///</summary>
    IsNullProperty : function()
    {
        var date = new NhsDate();
        // default datetype is exact
        Assert.IsTrue(!date.get_isNull());
        date.set_dateType(DateType.Null);
        Assert.IsTrue(date.get_isNull());
    },

    /// <summary>
    /// Test the ArgumentNullException for IsAddInstruction method
    ///</summary>
//    [ExpectedException(typeof(System.ArgumentNullException), "Should throw Null argument exception from IsAddInstruction method")]
    NullIsAddInstructionTest : function()
    {
        NhsDate.isAddInstruction(null);
    }
}

NhsDateTest.prototype.NullInstructionAddTest.expectedException={name:"Sys.ArgumentNullException", 
                                                         message:"Should throw Null argument exception from Add method"};
NhsDateTest.prototype.NullSourceDateAddTest.expectedException={name:"Sys.ArgumentNullException", 
                                                         message:"Should throw Null argument exception from Add method"};                                                         
NhsDateTest.prototype.InvalidInstructionAddTest.expectedException={name:"Sys.ArgumentOutOfRangeException", 
                                                         message:"Should throw ArgumentOutOfRangeException from Add method"};                                                          
NhsDateTest.prototype.SourceDateYearAddTest.expectedException={name:"Sys.ArgumentOutOfRangeException", 
                                                         message:"Should throw ArgumentOutOfRangeException from Add method"};                                                            
NhsDateTest.prototype.NullIsAddInstructionTest.expectedException={name:"Sys.ArgumentNullException", 
                                                         message:"Should throw Null argument exception from IsAddInstruction method"};                                                          
                                                         
                                                         
                                                         

File: Solutions/Main/NhsCui.Toolkit.Test/DateTest.cs
//-----------------------------------------------------------------------
// <copyright file="DateTest.cs" company="Microsoft Corporation and Crown copyright 2007 - 2010.">
// Copyright (c) Microsoft Corporation and Crown copyright 2007 - 2010.
// All rights reserved.
//
// CERTAIN PARTS OF THIS WORK CONTAIN SOFTWARE CODE THAT IS LICENSED 
// FOR USE UNDER THE MICROSOFT PUBLIC LICENSE. DISTRIBUTION, IN SOURCE CODE 
// OR OBJECT CODE FORM, OF THOSE PARTS MUST COMPLY WITH THE TERMS OF THE 
// PUBLIC LICENSE. SEE http://www.microsoft.com/opensource/licenses.mspx 
// FOR DETAILS.  
// IF YOU BRING A PATENT CLAIM AGAINST ANY CONTRIBUTOR OVER PATENTS THAT 
// YOU CLAIM ARE INFRINGED BY THE PUBLIC LICENSE SOFTWARE, YOUR PATENT 
// LICENSE FROM SUCH CONTRIBUTOR TO THE SOFTWARE ENDS AUTOMATICALLY.
//
// THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY 
// KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
// PARTICULAR PURPOSE.
// </copyright>
// <date>03-Jan-2007</date>
// <summary>Unit tests to test the Date class in the NhsCui.Toolkit.DateAndTime namespace</summary>
//-----------------------------------------------------------------------
namespace NhsCui.Toolkit.Test
{
    using System;
    using System.Text;
    using System.Collections.Generic;
    using System.Globalization;

    using Microsoft.VisualStudio.TestTools.UnitTesting;

    using NhsCui.Toolkit.DateAndTime;

    /// <summary>
    /// Unit tests to test the Date class in the NhsCui.Toolkit.DateAndTime namespace
    /// </summary>
    [TestClass]
    public class DateTest
    {
        /// <summary>
        /// Default ctor 
        /// </summary>
        public DateTest()
        {
        }

        /// <summary>
        /// Test ToString on DateType Exact
        /// </summary>
        [TestMethod]
        public void ConstructNhsDateDateTimeAndBool()
        {
            DateTime baseDateTime = DateTime.Now;

            NhsDate date = new NhsDate(baseDateTime, false);

            Assert.IsTrue(date.DateType == DateType.Exact);
            Assert.IsTrue(date.DateValue == baseDateTime);

            NhsDate date2 = new NhsDate(baseDateTime, true);

            Assert.IsTrue(date2.DateType == DateType.Approximate);
            Assert.IsTrue(date2.DateValue == baseDateTime);
        }

        #region Tests for Add(NhsDate sourceDate, string instruction) function
        
        /// <summary>
        /// Test the ArgumentNullException for instruction parameter
        /// </summary>
        [TestMethod()]
        [ExpectedException(typeof(System.ArgumentNullException), "Should throw Null argument exception from Add method")]
        public void NullInstructionAddTest()
        {
            DateTime nowAsItWasAtTheStartofTheTest = DateTime.Now;
            NhsDate date = new NhsDate(nowAsItWasAtTheStartofTheTest);
            NhsDate.Add(date, null);            
        }

        /// <summary>
        /// Test the ArgumentNullException for sourcedate parameter
        /// </summary>
        [TestMethod()]
        [ExpectedException(typeof(System.ArgumentNullException), "Should throw Null argument exception from Add method")]
        public void NullSourceDateAddTest()
        {
            NhsDate.Add(null, string.Format(CultureInfo.CurrentCulture, "+1{0}", NhsCui.Toolkit.Test.NhsCui_Toolkit_DateAndTime_Resources_NhsDateResourcesAccessor.DaysUnit));
        }
        
        /// <summary>
        /// Test the ArgumentException for instruction parameter
        /// </summary>
        [TestMethod()]
        [ExpectedException(typeof(System.ArgumentException), "Should throw ArgumentOutOfRangeException from Add method")]
        public void InvalidInstructionAddTest()
        {
            DateTime date = DateTime.Now;
            NhsDate sourcedate = new NhsDate(date);
            NhsDate.Add(sourcedate, string.Format(CultureInfo.CurrentCulture, "+1{0}", NhsCui.Toolkit.Test.NhsCui_Toolkit_DateAndTime_Resources_NhsDateResourcesAccessor.resourceCulture));
        }     

        /// <summary>
        /// Test the ArgumentException for sourcedate parameter being a year
        /// </summary>
        [TestMethod()]
        [ExpectedException(typeof(System.ArgumentException), "Should throw ArgumentOutOfRangeException from Add method")]
        public void SourceDateYearAddTest()
        {
            int year = 2007;
            NhsDate date = new NhsDate(year);
            NhsDate.Add(date, string.Format(CultureInfo.CurrentCulture, "+1{0}", NhsCui.Toolkit.Test.NhsCui_Toolkit_DateAndTime_Resources_NhsDateResourcesAccessor.DaysUnit));
        }

        /// <summary>
        /// Test the ability to do simple arithemtic updates to date by asking Date to add 1 day
        /// </summary>
        [TestMethod]
        public void DataAddDay()
        {
            DateTime nowAsItWasAtTheStartofTheTest = DateTime.Now; 

            NhsDate date = new NhsDate(nowAsItWasAtTheStartofTheTest);

            date.Add(string.Format(CultureInfo.CurrentCulture, "+1{0}", NhsCui.Toolkit.Test.NhsCui_Toolkit_DateAndTime_Resources_NhsDateResourcesAccessor.DaysUnit));

            Assert.AreEqual(date.DateValue, nowAsItWasAtTheStartofTheTest.AddDays(1));
        }

        /// <summary>
        /// Test the ability to do complex arithemtic updates to date by asking Date to add 1 month and 1 day
        /// </summary>
        [TestMethod]
        public void DataAddMonthAndDay()
        {
            DateTime nowAsItWasAtTheStartofTheTest = DateTime.Now; 

            NhsDate date = new NhsDate(nowAsItWasAtTheStartofTheTest);

            date.Add(string.Format(CultureInfo.CurrentCulture, "+1{0}+1{1}", NhsCui.Toolkit.Test.NhsCui_Toolkit_DateAndTime_Resources_NhsDateResourcesAccessor.MonthsUnit, NhsCui.Toolkit.Test.NhsCui_Toolkit_DateAndTime_Resources_NhsDateResourcesAccessor.DaysUnit));

            Assert.AreEqual(date.DateValue, nowAsItWasAtTheStartofTheTest.AddMonths(1).AddDays(1));
        }

        /// <summary>
        /// Test the ability to do simple arithemtic updates to date by asking Date to Add 7 years
        /// </summary>
        [TestMethod]
        public void DataAdd7Years()
        {
            DateTime nowAsItWasAtTheStartofTheTest = DateTime.Now; 

            NhsDate date = new NhsDate(nowAsItWasAtTheStartofTheTest);

            date.Add(string.Format(CultureInfo.CurrentCulture, "+7{0}", NhsCui.Toolkit.Test.NhsCui_Toolkit_DateAndTime_Resources_NhsDateResourcesAccessor.YearsUnit));
            
            Assert.AreEqual(nowAsItWasAtTheStartofTheTest.AddYears(7), date.DateValue);
        }

        /// <summary>
        /// Test the ability to do simple arithemtic updates to date by asking Date to Subtract 7 years
        /// </summary>
        [TestMethod]
        public void DataSubtract7Years()
        {
            DateTime nowAsItWasAtTheStartofTheTest = DateTime.Now; 

            NhsDate date = new NhsDate(nowAsItWasAtTheStartofTheTest);

            date.Add(string.Format(CultureInfo.CurrentCulture, "-7{0}", NhsCui.Toolkit.Test.NhsCui_Toolkit_DateAndTime_Resources_NhsDateResourcesAccessor.YearsUnit));

            Assert.AreEqual(nowAsItWasAtTheStartofTheTest.AddYears(-7), date.DateValue);
        }

        /// <summary>
        /// Test the ability to do complex arithemtic updates to date by asking Date to Subtract 2 years AND add 5 days
        /// </summary>
        [TestMethod]
        public void DataSubtract2YearsAdd5Days()
        {
            DateTime nowAsItWasAtTheStartofTheTest = DateTime.Now; 

            NhsDate date = new NhsDate(nowAsItWasAtTheStartofTheTest);

            date.Add(string.Format(CultureInfo.CurrentCulture, "-2{0}+5{1}", NhsCui.Toolkit.Test.NhsCui_Toolkit_DateAndTime_Resources_NhsDateResourcesAccessor.YearsUnit, NhsCui.Toolkit.Test.NhsCui_Toolkit_DateAndTime_Resources_NhsDateResourcesAccessor.DaysUnit));

            Assert.AreEqual(nowAsItWasAtTheStartofTheTest.AddYears(-2).AddDays(5), date.DateValue);
        }

        #endregion    

        /// <summary>
        /// Test ToString on DateType Exact
        /// </summary>
        [TestMethod]
        public void ToStringExact()
        {
            NhsDate date = new NhsDate(new System.DateTime(1974, 3, 26));

            Assert.AreEqual("26-Mar-1974", date.ToString());

            Assert.AreEqual("26-Mar-1974", date.ToString(false), "Check that default for 'include day of week' flag is correct");

            // try with includeDayOfWeek flag
            Assert.AreEqual("Tue 26-Mar-1974", date.ToString(true));
        }

        /// <summary>
        /// Test ToString on DateType Exact
        /// </summary>
        [TestMethod]
        public void ToStringExactGetRelativeText()
        {
            NhsDate date = new NhsDate(System.DateTime.Today);

            GlobalizationService gc = new GlobalizationService();

            // Today works
            Assert.AreEqual(NhsCui.Toolkit.Test.NhsCui_Toolkit_DateAndTime_Resources_NhsDateResourcesAccessor.Today, date.ToString(false, false, true, CultureInfo.CurrentCulture), "RelativeText of Today is not working");

            date.DateValue = DateTime.Today.AddDays(1);

            Assert.AreEqual(NhsCui.Toolkit.Test.NhsCui_Toolkit_DateAndTime_Resources_NhsDateResourcesAccessor.Tomorrow, date.ToString(false, false, true, CultureInfo.CurrentCulture), "RelativeText of Tomorrow is not working");

            date.DateValue = DateTime.Today.AddDays(-1);

            Assert.AreEqual(NhsCui.Toolkit.Test.NhsCui_Toolkit_DateAndTime_Resources_NhsDateResourcesAccessor.Yesterday, date.ToString(false, false, true, CultureInfo.CurrentCulture), "RelativeText of Yesterday is not working");
            
            // Check the "Who Wins condition when "Show Day of Week" and "Show Relative Text" are both true
            Assert.AreEqual(NhsCui.Toolkit.Test.NhsCui_Toolkit_DateAndTime_Resources_NhsDateResourcesAccessor.Yesterday, date.ToString(true, false, true, CultureInfo.CurrentCulture), "When includeDayOfWeek, showRelativeText are both true, showRelativeText should win. It is not");

            DateTime testDate = new DateTime(1974, 3, 26);

            date.DateValue = testDate;

            Assert.AreEqual(testDate.ToString(gc.ShortDatePattern, CultureInfo.CurrentCulture), date.ToString(), "Check that default for 'include day of week' flag is correct");

            // try with includeDayOfWeek flag

            Assert.AreEqual(testDate.ToString(gc.ShortDatePatternWithDayOfWeek, CultureInfo.CurrentCulture), date.ToString(true), "Check that default for 'include day of week' flag is correct");
        }

        /// <summary>
        /// Test ToString on DateType Year
        /// </summary>
        [TestMethod]
        public void ToStringYear()
        {
            System.DateTime baseDateTime = System.DateTime.Now;

            NhsDate date = new NhsDate(baseDateTime);

            date.DateType = DateType.Year;

            Assert.AreEqual("0000", date.ToString(), "Check the default of Year when an explicit date is passed to ctor");

            date = new NhsDate(baseDateTime.Year);

            date.DateType = DateType.Year;

            Assert.AreEqual(baseDateTime.Year.ToString(CultureInfo.CurrentCulture), date.ToString());
        }

        /// <summary>
        /// Test ToString on DateType YearMonth
        /// </summary>
        [TestMethod]
        public void ToStringYearMonth()
        {
            System.DateTime baseDateTime = System.DateTime.Now;

            NhsDate date = new NhsDate(baseDateTime);

            date.DateType = DateType.YearMonth;

            Assert.AreEqual(string.Format(CultureInfo.InvariantCulture, "{0}-{1}", CultureInfo.CurrentUICulture.DateTimeFormat.MonthNames[1 - 1], "0000"), date.ToString(), "Check the default of Month and Year when an explicit date is passed to ctor");
            
            // Now set the date to a year and a month which is what DateType = YearMonth prefers
            date = new NhsDate(1974, 3);

            Assert.AreEqual(
                string.Format(CultureInfo.InvariantCulture, "{0}-{1}", CultureInfo.CurrentUICulture.DateTimeFormat.MonthNames[3 - 1], "1974"), date.ToString());
        }

        /// <summary>
        /// Test ToString on DateType Null
        /// </summary>
        [TestMethod]
        public void ToStringNull()
        {
            NhsDate date = new NhsDate();
            date.DateType = DateType.Null;
            Assert.IsTrue(date.ToString().Length == 0);
        }    

        /// <summary>
        /// Test Parse for a Month and Year
        /// </summary>
        [TestMethod]
        public void ParseMonthAndYear()
        {
            System.DateTime baseDateTime = System.DateTime.Now;

            NhsDate date = new NhsDate(baseDateTime);

            date.DateType = DateType.YearMonth;
            date.Year = 1974;
            date.Month = 3;

            NhsDate parsedDate;

            Assert.IsTrue(NhsDate.TryParseExact("jan-2008", out parsedDate, CultureInfo.InvariantCulture), "Try parse failed");
            Assert.AreEqual(parsedDate.Month, 1, "Month not being parsed");
            Assert.AreEqual(parsedDate.Year, 2008, "Month not being parsed");

            Assert.IsTrue(NhsDate.TryParseExact("january-2008", out parsedDate, CultureInfo.InvariantCulture), "Try parse failed");
            Assert.AreEqual(parsedDate.Month, 1, "Month not being parsed");
            Assert.AreEqual(parsedDate.Year, 2008, "Month not being parsed");

            Assert.IsTrue(NhsDate.TryParseExact("1-2008", out parsedDate, CultureInfo.InvariantCulture), "Try parse failed");
            Assert.AreEqual(parsedDate.Month, 1, "Month not being parsed");
            Assert.AreEqual(parsedDate.Year, 2008, "Month not being parsed");

            Assert.IsTrue(NhsDate.TryParseExact("10-2008", out parsedDate, CultureInfo.InvariantCulture), "Try parse failed");
            Assert.AreEqual(parsedDate.Month, 10, "Month not being parsed");
            Assert.AreEqual(parsedDate.Year, 2008, "Month not being parsed");

            Assert.IsTrue(NhsDate.TryParseExact(date.ToString(), out parsedDate, CultureInfo.InvariantCulture), string.Format(CultureInfo.InvariantCulture, "TryParse failed with date, {0}", date.ToString()));

            Assert.AreEqual(date.Month, parsedDate.Month, "Parse did not error but Month does not match");
            Assert.AreEqual(date.DateType, parsedDate.DateType, "Parse did not error but DateType does not match");
            Assert.AreEqual(date.Year, parsedDate.Year, "Parse did not error but Year does not match");
        }

        /// <summary>
        /// Test Parse for a Null
        /// </summary>
        [TestMethod]
        public void ParseNullIndex()
        {
            NhsDate date = new NhsDate();

            date.DateType = DateType.NullIndex;
            date.NullIndex = 14;
            
            // Muddy the water by assigning Month and year
            date.Year = 2007;
            date.Month = 3;

            NhsDate parsedDate;

            Assert.IsTrue(NhsDate.TryParseExact(date.ToString(), out parsedDate, CultureInfo.InvariantCulture), string.Format(CultureInfo.InvariantCulture, "TryParse failed with date, {0}", date.ToString()));

            Assert.AreEqual(date.DateType, parsedDate.DateType, "Parse did not error but DateType does not match");
            Assert.AreEqual(date.NullIndex, parsedDate.NullIndex, "Parse did not error but NullIndex does not match");
        }

        /// <summary>
        /// Test Parse for an Invalid NullIndex
        /// </summary>
        [TestMethod]
        public void ParseInvalidNullIndex()
        {
            NhsDate parsedDate;

            string invalidNullIndexString = "Null:-256";

            Assert.IsFalse(NhsDate.TryParseExact(invalidNullIndexString, out parsedDate, CultureInfo.InvariantCulture), string.Format(CultureInfo.InvariantCulture, "TryParse failed to return false with string of, {0}", invalidNullIndexString));

            invalidNullIndexString = "Null:124";

            Assert.IsFalse(NhsDate.TryParseExact(invalidNullIndexString, out parsedDate, CultureInfo.InvariantCulture), string.Format(CultureInfo.InvariantCulture, "TryParse failed to return false with string of, {0}", invalidNullIndexString));
        }

        /// <summary>
        /// Test Parse for a free text
        /// </summary>
        [TestMethod]
        public void ParseDateText()
        {
            NhsDate date = new NhsDate("26-March-1974");

            NhsDate parsedDate;

            Assert.IsTrue(NhsDate.TryParseExact(date.ToString(), out parsedDate, CultureInfo.InvariantCulture), string.Format(CultureInfo.InvariantCulture, "TryParse failed with date, {0}", date.ToString()));

            Assert.AreEqual(date.DateType, parsedDate.DateType, "Parse did not error but DateType does not match");
            Assert.AreEqual(date.DateValue, parsedDate.DateValue, "Parse did not error but DateValue does not match");
        }

        /// <summary>
        /// Test Parse null
        ///</summary>
        [TestMethod()]
        public void ParseNull()
        {
            NhsDate date = NhsDate.ParseExact(null, CultureInfo.CurrentCulture);
            Assert.AreEqual<DateType>(date.DateType, DateType.Null, "Parse did not error but DateType does not match");
        }

        /// <summary>
        /// Test Parse empty string
        ///</summary>
        [TestMethod()]
        public void ParseEmptyString()
        {
            NhsDate date = NhsDate.ParseExact(string.Empty, CultureInfo.CurrentCulture);
            Assert.AreEqual<DateType>(date.DateType, DateType.Null, "Parse did not error but DateType does not match");
        }

        /// <summary>
        /// Test tryParse to see thatit gets a False when it hsould
        ///</summary>
        [TestMethod()]
        public void TryParseFailures()
        {
            NhsDate result;

            Assert.IsFalse(NhsDate.TryParseExact("31-12", out result, CultureInfo.CurrentCulture));

            Assert.IsFalse(NhsDate.TryParseExact("01-2008", out result, CultureInfo.CurrentCulture));

            Assert.IsFalse(NhsDate.TryParseExact("-12", out result, CultureInfo.CurrentCulture));
        }

        /// <summary>
        /// Test IsNull Property
        ///</summary>
        [TestMethod()]
        public void IsNullProperty()
        {
            NhsDate date = new NhsDate();

            // default datetype is exact
            Assert.IsFalse(date.IsNull);
            date.DateType = DateType.Null;
            Assert.IsTrue(date.IsNull);
        }

        /// <summary>
        /// Test the ArgumentNullException for IsAddInstruction method
        ///</summary>
        [TestMethod()]
        [ExpectedException(typeof(System.ArgumentNullException), "Should throw Null argument exception from IsAddInstruction method")]
        public void NullIsAddInstructionTest()
        {
            NhsDate.IsAddValid(null);
        }
    }
}
